Meta Checkout URL Implementation

sol-loup opened the pull request:
# Meta Checkout URL Implementation

This PR adds support for Meta's checkout URL specification, allowing direct linking to the checkout page with pre-populated cart items and coupon codes.

## What's Changed

- Added `AddToCartLinkV1` controller that processes URL parameters to populate the cart
- Added layout XML file to render the checkout page with consistent user experience
- Implemented support for multiple products and coupon codes via URL parameters

## How It Works

The controller accepts URL parameters in this format:
/checkout/cart/addtocartlinkv1?products=123:2,456:1&coupon=SAVE10

![Screenshot 2025-02-27 at 2 45 27 PM](https://github.com/user-attachments/assets/57691f4b-38ce-43f6-ba61-460da11cdc77)



This will:
1. Clear the existing cart (per Meta requirements)
2. Add 2 units of product #123 and 1 unit of product #456
3. Apply the "SAVE10" coupon code
4. Display the checkout page

The implementation handles error cases gracefully, showing appropriate messages for invalid products or coupons.


### Resolved issues:
1. [x] resolves magento/magento2#39700: Meta Checkout URL Implementation

sol-loup committed:
Adding AddToCartLinkV1 controller and checkout_cart_link route for Meta Checkout URL implementation

engcom-Hotel committed:
Merge branch '2.4-develop' into 2.4-develop

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
@magento create issue

engcom-Hotel committed:
Merge branch '2.4-develop' into 2.4-develop

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,196 @@
+/**
+ * Controller for Meta Checkout URL implementation
+ */
+class AddToCartLinkV1 implements HttpGetActionInterface, ActionInterface
```
Remove the redundant implementation of `ActionInterface`. It is already implemented by HttpGetActionInterface.

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,196 @@
+        CouponFactory $couponFactory,
+        Usage $couponUsage,
+        ManagerInterface $messageManager
+    ) {
```
Use [Constructor readonly properties](https://php.watch/versions/8.1/readonly#constructor-readonly)

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,196 @@
+                    $qty = $item['qty'];
+                    
+                    $product = $this->productRepository->getById($productId);
+                    $this->cart->addProduct($product, ['qty' => $qty]);
```
I guess we can pass `$product->getId()`, instead of passing `$product` as `addProduct` method requires either `int` or `\Magento\Catalog\Model\Product` parameter

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,196 @@
+     *
+     * @return \Magento\Framework\Controller\ResultInterface
+     */
+    public function execute()
```
Add a return type declaration.

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,196 @@
+     * @param string $productsParam
+     * @return array
+     */
+    private function parseProductsParam($productsParam)
```
Add a return type declaration and also add a parameter type declaration.

engcom-Hotel commented on app/code/Magento/Checkout/view/frontend/layout/checkout_cart_addtocartlinkv1.xml:
```diff
@@ -0,0 +1,11 @@
+ * See COPYING.txt for license details.
+ */
+-->
+<?xml version="1.0"?>
```
I think we can remove this redundant statement

engcom-Hotel reviewed (changes requested):
Hello @sol-loup,

Thanks for the contribution. 

I can see we can handle Simple product scenarios with this, but how can we handle the configurable products.

Also, I guess we can add an automated test for this newly added functionality.

engcom-Charlie commented:
Hi @sol-loup,

Thank you for your contribution!

Can you please look into the review comments and do the needful.

sol-loup committed:
addressing comments

sol-loup committed:
Merge branch '2.4-develop' of https://github.com/sol-loup/magento2 into 2.4-develop

sol-loup commented:
@engcom-Hotel @engcom-Charlie , I've made the requested updates. 

I've also added a unit test as requested; test output: 

vendor/bin/phpunit -c dev/tests/unit/phpunit.xml.dist vendor/magento/module-checkout/Test/Unit/Controller/Cart/.
PHPUnit 9.6.20 by Sebastian Bergmann and contributors.

.......                                                                                       7 / 7 (100%)

Time: 00:00.563, Memory: 10.00 MB

OK (7 tests, 47 assertions)

sol-loup commented:
> Hello @sol-loup,
> 
> Thanks for the contribution.
> 
> I can see we can handle Simple product scenarios with this, but how can we handle the configurable products.
> 
> Also, I guess we can add an automated test for this newly added functionality.

This is a good callout; I've adjusted the logic to prioritize sku instead. 

This will allow specific configurations to be purchased via SKU -- this will allow most standard configurable variable items to flow through this feature. Customizable products are a special case, and IMO, not well handled even by other e-commerce platforms. 

Example here: 
https://one.staging.magento-commerce-extension-testing.com/checkout/cart/addtocartlinkv1?products=12345:2

sol-loup committed:
Shifting to a SKU based approach

engcom-Charlie commented:
Hi @sol-loup,

Thank you! I can see that you have taken care all the review comments except https://github.com/magento/magento2/pull/39667#discussion_r1983183263. Are you going to finish this one too or should I check it further.

sol-loup committed:
Addressing PR comments

engcom-Charlie commented:
> > Hi @sol-loup,
> > Thank you! I can see that you have taken care all the review comments except [#39667 (comment)](https://github.com/magento/magento2/pull/39667#discussion_r1983183263). Are you going to finish this one too or should I check it further.
> 
> Apologies for the oversight; I've implemented the string-id based check, and the PR is now ready for another review from your team

Thank you @sol-loup!

No Problem, thank you for the update! 

As all the review comments have been taken care, moving this PR to review.

engcom-Hotel committed:
Merge branch '2.4-develop' into 2.4-develop

engcom-Hotel commented:
@magento run all tests

sol-loup committed:
Fixing lint complaints

sol-loup committed:
Merge branch '2.4-develop' of https://github.com/sol-loup/magento2 into 2.4-develop

sol-loup commented:
@engcom-Hotel  -- after reviewing the allure results, I have proactively resolved the majority of the lint complaints. However, I am seeing several complaints about copyright being missing or malformatted. I am copying the adobe copyright from other files. Can you advise?

sol-loup commented:
@magento run all tests

sol-loup committed:
Adjusting Adobe Copyright Comment

sol-loup commented:
> Sure @sol-loup,
> 
> To resolve the copyright issue you can refer to this file
> 
> https://github.com/magento/magento2/blob/727294d82ecd43bb36588df38130e2c5ee08bc2f/lib/internal/Magento/Framework/Api/SearchCriteria/CollectionProcessor/JoinProcessor.php#L2-L5
> 
> .
> As you can see it contains the copyright information as below:
> 
> ```
> /**
>  * Copyright [year code created] Adobe 
>  * All Rights Reserved.
> */
> ```
> 
> So please update the information as below. Moving back to `Changes Requested` as some automated tests are still failing.
> 
> Thanks

Thanks @engcom-Hotel . I've resolved the remaining static lint issues. Looking at the functional testing failures, they don't appear to be related to my changes?

engcom-Hotel committed:
Merge branch '2.4-develop' into 2.4-develop

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,174 @@
+     */
+    public function __construct(
+        Context $context,
+        private readonly CheckoutSession $checkoutSession,
```
`$checkoutSession` property is never used, I guess we can remove this.

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,174 @@
+        private readonly ProductRepositoryInterface $productRepository,
+        private readonly Cart $cart,
+        private readonly PageFactory $resultPageFactory,
+        private readonly RedirectFactory $resultRedirectFactory,
```
These properties is also never used, I guess we can remove this.

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,174 @@
+        Context $context,
+        private readonly CheckoutSession $checkoutSession,
+        private readonly ProductRepositoryInterface $productRepository,
+        private readonly Cart $cart,
```
Class `Cart` has been deprecated, please replace it with `\Magento\Quote\Model\Quote`

engcom-Hotel reviewed (changes requested):
Hello @sol-loup,

Thanks for the making the changes!

I can see some static test are still failing. Other failures seems flaky to me. Also please address some review comment below and make the adjustments accordingly in the test file as well.

Thanks

sol-loup committed:
Changing implementation to use quote

engcom-Hotel commented:
@magento run all tests

engcom-Hotel reviewed (approved):
Thanks @sol-loup, for making the changes. I am approving the PR.

engcom-Bravo committed:
Merge branch '2.4-develop' into 2.4-develop

engcom-Bravo commented:
@magento run all tests

engcom-Bravo commented:
Hi @sol-loup,

Thanks for the collaboration & contribution!

### :heavy_check_mark: QA Passed


**Preconditions:**

- Install fresh Magento **`2.4-develop`**


**Before:** :heavy_multiplication_x: 



<img width="1436" alt="Screenshot 2025-04-02 at 2 25 34 pm" src="https://github.com/user-attachments/assets/c1a76518-a93e-496b-9981-6f95cb398a96" />




**After:** :heavy_check_mark: 



<img width="1436" alt="Screenshot 2025-04-02 at 2 35 26 pm" src="https://github.com/user-attachments/assets/72f2db3b-4807-44f4-8280-9719823ec8f7" />




Builds are failed. Hence, moving this PR to **Extended Testing**.

Thanks.

engcom-Charlie commented:
@magento run Functional Tests B2B

engcom-Charlie commented:
@magento run Functional Tests B2B

engcom-Charlie commented:
One of the Functional B2B test failure is not consistent in recent 2 builds and the other one is has not failed with the same error. Hence moving this PR to Merge in Progress.

**Run 1:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39667/aed35d01ff8d5224b69af4ac290e4090/Functional/allure-report-b2b/index.html#categories/ba729ce87e7059ef4466d4300850777a/8a2694d36b61e5d6/
<img width="1105" alt="image" src="https://github.com/user-attachments/assets/92370529-29ce-4cf0-ac7a-de2055619313" />

**Run 2:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39667/657390fbde4ccdce471439b58622c1eb/Functional/allure-report-b2b/index.html#categories

<img width="1081" alt="image" src="https://github.com/user-attachments/assets/8632b7d1-c7f5-4b69-b53c-7ab86f4abf5e" />

engcom-Charlie commented:
@engcom-Hotel I guess it will be good if we make this feature configurable one. Moving it to Pending Review again.

engcom-Hotel reviewed (changes requested):
Hello @sol-loup,

Thank you for making the changes based on the review comments!

Following our internal discussion on this feature, we believe it would be beneficial to add an option to enable or disable the feature from the admin panel.

Could you please make the necessary adjustments at your earliest convenience?

Thanks

sol-loup commented:
> Hi @sol-loup and @engcom-Bravo,
> 
> I have tried to check different scenarios on my local environment on PR branch When I have tried to browse http://39667.local/checkout/cart/addtocartlinkv1?products=abc:5, it gave me correct page.
> 
> <img alt="image" width="1208" src="https://private-user-images.githubusercontent.com/51680850/430385304-973b9a65-a9bf-467a-9281-5d7b356cff1c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwNDgyNTUsIm5iZiI6MTc0NDA0Nzk1NSwicGF0aCI6Ii81MTY4MDg1MC80MzAzODUzMDQtOTczYjlhNjUtYTliZi00NjdhLTkyODEtNWQ3YjM1NmNmZjFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA3VDE3NDU1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU5NzhhMzcwNmU1NjEyN2Y0MzZmYzIyZmIyYjhmYmE1MmU3YjkxOWI2NmYwZjkzYjEyY2Q2ZjliODc4ODAxZWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gPMu1bBJ-SxCpI-u9i5wjlM69M3B9huh81TUybvbemo">
> But when I have tweaked the URL by adding extra letter before product, like http://39667.local/checkout/cart/addtocartlinkv1?aproducts=abc:5&b=23 then its giving me checkout page but without the item details. I think, we should handle all such scenarios here. <img alt="image" width="1234" src="https://private-user-images.githubusercontent.com/51680850/430387242-a30ae667-47b7-43a0-b9ff-66addd4a8d64.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwNDgyNTUsIm5iZiI6MTc0NDA0Nzk1NSwicGF0aCI6Ii81MTY4MDg1MC80MzAzODcyNDItYTMwYWU2NjctNDdiNy00M2EwLWI5ZmYtNjZhZGRkNGE4ZDY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA3VDE3NDU1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQwYzc5MGM4NWJhMTI0MWE0YTI5YTgwNjY4NDBmNDJiNjNhYWNmZjA4MzliMTQ2MzU5NGJjYjA0NDcyY2NlMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iJoFqliDiYQbgruqrQ_pMckPByUvIy2fCtEdZcxi7UA">
> 
> @sol-loup let me know if I am missing anything. @engcom-Bravo please check it with all the possible scenarios. Till then moving it to Ready for Testing again.

Hey @engcom-Charlie. In this case, you are passing an invalid parameter, and you are just landing on your previous valid cart, whatever that cart may have been. If you feel an explicit error message is more appropriate here, happy to add that.

engcom-Charlie commented:
> > Hi @sol-loup and @engcom-Bravo,
> > I have tried to check different scenarios on my local environment on PR branch When I have tried to browse http://39667.local/checkout/cart/addtocartlinkv1?products=abc:5, it gave me correct page.
> > <img alt="image" width="1208" src="https://private-user-images.githubusercontent.com/51680850/430385304-973b9a65-a9bf-467a-9281-5d7b356cff1c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwNDgyNTUsIm5iZiI6MTc0NDA0Nzk1NSwicGF0aCI6Ii81MTY4MDg1MC80MzAzODUzMDQtOTczYjlhNjUtYTliZi00NjdhLTkyODEtNWQ3YjM1NmNmZjFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA3VDE3NDU1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU5NzhhMzcwNmU1NjEyN2Y0MzZmYzIyZmIyYjhmYmE1MmU3YjkxOWI2NmYwZjkzYjEyY2Q2ZjliODc4ODAxZWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gPMu1bBJ-SxCpI-u9i5wjlM69M3B9huh81TUybvbemo">
> > But when I have tweaked the URL by adding extra letter before product, like http://39667.local/checkout/cart/addtocartlinkv1?aproducts=abc:5&b=23 then its giving me checkout page but without the item details. I think, we should handle all such scenarios here. <img alt="image" width="1234" src="https://private-user-images.githubusercontent.com/51680850/430387242-a30ae667-47b7-43a0-b9ff-66addd4a8d64.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwNDgyNTUsIm5iZiI6MTc0NDA0Nzk1NSwicGF0aCI6Ii81MTY4MDg1MC80MzAzODcyNDItYTMwYWU2NjctNDdiNy00M2EwLWI5ZmYtNjZhZGRkNGE4ZDY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA3VDE3NDU1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQwYzc5MGM4NWJhMTI0MWE0YTI5YTgwNjY4NDBmNDJiNjNhYWNmZjA4MzliMTQ2MzU5NGJjYjA0NDcyY2NlMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iJoFqliDiYQbgruqrQ_pMckPByUvIy2fCtEdZcxi7UA">
> > @sol-loup let me know if I am missing anything. @engcom-Bravo please check it with all the possible scenarios. Till then moving it to Ready for Testing again.
> 
> Hey @engcom-Charlie. In this case, you are passing an invalid parameter, and you are just landing on your previous valid cart, whatever that cart may have been. If you feel an explicit error message is more appropriate here, happy to add that.

Hi @sol-loup,

Yes, I have tweaked the URL by passing wrong params. Showing the previous cart on this wrong URL doesn't make any sense. It will be good if we handled it correctly. You can check how such cases are getting handled right now. Eg. either by showing error or most appropriately by showing below page. 

<img width="1223" alt="image" src="https://github.com/user-attachments/assets/19f5f2b5-c647-448e-bbc8-fe1be34de592" />

sol-loup committed:
adding configuration values to enable and disable add to cart linking

sol-loup committed:
Merge branch '2.4-develop' of https://github.com/sol-loup/magento2 into 2.4-develop

sol-loup commented:
Hey @engcom-Charlie -- I have applied the change allowing for configuration from the "Checkout" configurations in Magento: 
![Screenshot 2025-04-16 at 3 18 17 PM](https://github.com/user-attachments/assets/c60ad19b-c16c-456c-b4e3-181cbc0cbb37)

I have also added logic adding a &store param, allowing sellers on multi-site setups to specify which store (if any) they'd like the checkout to occur from. 

I have added the noroute behavior in the case where the functionality is disabled. However, I am not comfortable proceeding with the cart deletion logic you have suggested here. This could potentially be a destructive act if the buyer has an active cart session from some other previous shopping session. In the event the Add To Cart Link is invalid, I would prefer not to delete their pre-existing cart.

engcom-Charlie commented:
Hi @sol-loup,

Thank you for the updating.  @engcom-Hotel Since the review comments have been worked on, moving this PR into review.

engcom-Hotel committed:
Merge branch '2.4-develop' into 2.4-develop

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,280 @@
+        $couponCode = $this->_request->getParam('coupon', '');
+
+        // Get quote from checkout session (should now reflect the correct store)
+        $quote = $this->checkoutSession->getQuote();
```
Calling `getQuote` method may return exception, surround it with `try... catch` block.

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,280 @@
+                    $this->messageManager->addErrorMessage(
+                        __('The coupon code "%1" is not valid or cannot be applied.', $couponCode)
+                    );
+                } else {
```
This `else` statement has empty body, I think we can remove it.

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,280 @@
+        $quote = $this->checkoutSession->getQuote();
+
+        // Ensure quote is associated with the correct store ID after potential switch
+        if ($quote->getStoreId() !== $this->storeManager->getStore()->getId()) {
```
Handle `NoSuchEntityException` for `getStore` method

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,280 @@
+
+            // Save quote and collect totals
+            $quote->collectTotals();
+            $quote->save();
```
Handle `Exception` for `save`

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,280 @@
+        if (!empty($couponCode)) {
+            try {
+                // Ensure coupon is applied in the context of the potentially switched store
+                $quote->setCouponCode($couponCode)->collectTotals()->save();
```
The AbstractModel `save` method has been deprecated, hence please use resource model `save` method.

engcom-Hotel commented on app/code/Magento/Checkout/Test/Unit/Controller/Cart/AddToCartLinkV1Test.php:
```diff
@@ -0,0 +1,342 @@
+        // Create Quote mock with all required methods
+        $this->_quoteMock = $this->getMockBuilder(Quote::class)
+            ->onlyMethods(['removeAllItems', 'addProduct', 'collectTotals', 'save'])
+            ->addMethods(['setCouponCode', 'getCouponCode'])
```
`addMethods` is deprecated, please see https://github.com/sebastianbergmann/phpunit/issues/5320

engcom-Hotel commented on app/code/Magento/Checkout/Test/Unit/Controller/Cart/AddToCartLinkV1Test.php:
```diff
@@ -0,0 +1,342 @@
+use Magento\Framework\App\RequestInterface;
+use Magento\Framework\Controller\ResultInterface;
+use Magento\Framework\Message\ManagerInterface;
+use Magento\Framework\TestFramework\Unit\Helper\ObjectManager;
```
`ObjectManager` manager is deprecated, class under test should be instantiated with `new` keyword with explicit dependencies declaration

engcom-Hotel commented on app/code/Magento/Checkout/etc/config.xml:
```diff
@@ -21,6 +21,7 @@
                 <number_items_to_display_pager>20</number_items_to_display_pager>
                 <crosssell_enabled>1</crosssell_enabled>
                 <enable_clear_shopping_cart>0</enable_clear_shopping_cart>
+                <enable_add_to_cart_link_v1>1</enable_add_to_cart_link_v1>
```
The default value for the configuration is set to `1` (enabled). This might introduce unintended behavior for stores unaware of this new feature. So please disable it by default

engcom-Hotel commented on app/code/Magento/Checkout/Controller/Cart/AddToCartLinkV1.php:
```diff
@@ -0,0 +1,280 @@
+            $parts = explode(':', $pair);
+            if (count($parts) === 2) {
+                $identifier = trim($parts[0]);
+                $qty = filter_var($parts[1], FILTER_VALIDATE_INT); // Validate quantity is an integer
```
To prevent misuse, please add a maximum limit on the number of products that can be processed in one request.

engcom-Hotel reviewed (changes requested):
Hello @sol-loup,

Thanks for making the changes as per the review comment. 

Please fix the failed static and unit test failures and some review comments as below.

Functional test failures seems flaky to me and for semantic version checker failure we need to create an internal JIRA for approval.

Thanks

engcom-Charlie commented:
Hi @sol-loup,

Can you please have a look into the review comments as well as fix the build failures.

Static test failures: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39667/1fdd1741792c608604490da5fac3f74b/Statics/console-error-logs.html

Unit test failures: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39667/a3e18407c90d68f9ce142440a79d4183/Unit/console-error-logs.html

engcom-Charlie commented:
Hi @sol-loup,

Did you get a chance to look into [above](https://github.com/magento/magento2/pull/39667#issuecomment-2830504745) points.

engcom-Hotel commented:
Hello @sol-loup,

Regarding this https://github.com/magento/magento2/pull/39667#discussion_r2057801871, this is a completely new feature in Magento. Enabling it silently could lead to store owners being unaware of its presence or impact. That’s why I suggested keeping it disabled by default, allowing store owners to enable it explicitly if they choose to use the feature.

Thanks

sol-loup committed:
changes to satisfy feedback

sol-loup commented:
> Hello @sol-loup,
> 
> Regarding this [#39667 (comment)](https://github.com/magento/magento2/pull/39667#discussion_r2057801871), this is a completely new feature in Magento. Enabling it silently could lead to store owners being unaware of its presence or impact. That’s why I suggested keeping it disabled by default, allowing store owners to enable it explicitly if they choose to use the feature.
> 
> Thanks

@engcom-Hotel

I’d like to keep this flag enabled for a few reasons:

1. Accidental use is virtually impossible
   a. Path & params: ​/cart/addtocartlinkv1?productId={{id}}:{{qty}}&coupon={{code}} are not guessable. The handler only fires when that exact pattern is present, so routine crawlers, shoppers, or legacy integrations never hit it unintentionally.

2. Every modern platform already ships it ON
        - Shopify – “Cart permalinks” ([public docs](https://help.shopify.com/en/manual/products/details/cart-permalink)).
        - WooCommerce – “Add‑to‑cart URLs” ([public docs](https://woocommerce.com/document/quick-guide-to-woocommerce-add-to-cart-urls/)).
        - Merchants and growth‑teams now expect this behaviour to “just work.” Disabling it by default forces Magento stores to install extensions or patch core, adding friction that our competitors don’t impose.

3. Business upside > Theoretical downside
        - Campaign lift: agencies report higher CVR when ads land on a pre‑filled cart vs. PDP.
        - Ad‑platform parity: Meta, Google P‑Max, TikTok, and affiliate feeds all generate these links automatically for E-Comm platforms once the endpoint exists. Having it active out‑of‑the‑box means Magento merchants will benefit from the same link behaviors on these platforms without additional configuration.
4. Opt‑out remains: If a merchant truly doesn’t want the feature, they can disable it in Stores › Configuration › Sales › Checkout › “Enable Cart‑URL links”.

In my opinion, this change aligns Magento with industry checkout norms, removes needless adoption friction, and lets merchants start capturing higher‑intent traffic the moment they upgrade.

Happy to tweak wording or docs, but I believe default‑ON is the lowest‑friction, highest‑ROI path for the ecosystem.

sol-loup committed:
Merge branch '2.4-develop' of https://github.com/sol-loup/magento2 into 2.4-develop

engcom-Hotel commented:
Hello @sol-loup,

Thank you for your detailed explanation and for providing industry comparisons and business rationale. Let us discussing this feature internally, meanwhile moving this PR `On Hold`.

Thanks

engcom-Hotel commented:
Hello @sol-loup,

Thank you for your detailed explanation and for providing industry comparisons and business rationale. I appreciate the thoughtfulness behind your reasoning. However, we would like to reiterate some considerations and address your points according to discussion with the internal team:

**_Accidental Use_**: 

While the URL pattern is specific, we cannot entirely rule out the possibility of unintended activation, especially in edge cases or during system integrations. A disabled-by-default approach ensures that only merchants who fully understand and intend to use the feature will enable it.

**_Industry Standards_**: 

While Shopify and WooCommerce may have similar features enabled by default, Magento's ecosystem and user base have their own unique expectations and workflows. Silent enablement of new features could lead to confusion, especially for merchants upgrading from an older version who may not be aware of the change.

**_Business Benefits vs. Merchant Control_**: 

While the business upside is promising, we must prioritize giving merchants control over their stores' functionality. By keeping the feature disabled initially, we empower merchants to make a deliberate decision to enable it, ensuring they fully understand its behavior and implications.

**_Opt-Out Option_**: 

While it's true that merchants can disable the feature, it's generally better to follow an opt-in model for new functionality, especially when it changes how the platform behaves out of the box. This avoids surprises and ensures a consistent experience for existing users.

Given these considerations, I would still advocate for keeping the feature disabled by default upon release. This approach minimizes potential risks while still allowing merchants to leverage the feature if they see fit. I'd be happy to collaborate further to refine the documentation or messaging around this feature to ensure its value is clearly communicated.

Thanks

engcom-Hotel commented:
Hello @sol-loup,

Please update us on this PR. 

We are moving this PR to `Changes Requested`.

Thanks

engcom-Hotel commented:
Hello @sol-loup,

Gentle reminder for this PR.

Thanks

engcom-Charlie commented:
> Hey @engcom-Charlie -- I have applied the change allowing for configuration from the "Checkout" configurations in Magento: ![Screenshot 2025-04-16 at 3 18 17 PM](https://private-user-images.githubusercontent.com/129231008/434975630-c60ad19b-c16c-456c-b4e3-181cbc0cbb37.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA1MjIzOTUsIm5iZiI6MTc2MDUyMjA5NSwicGF0aCI6Ii8xMjkyMzEwMDgvNDM0OTc1NjMwLWM2MGFkMTliLWMxNmMtNDU2Yy1iNGUzLTE4MWNiYzBjYmIzNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAxNVQwOTU0NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yY2FiNGQ4MDA2MWJlNDZlMWIxNWY1NGUwM2M0NWFmMDdkYTAwMzBlYzcxZWY4YTM0ZDc4MDkyYTYzYzI5ZTEyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.JCLS7eeKx3mrN3cJfT7mgO3ZxZ91SNAZ0pvldrbYrHs)
> 
> I have also added logic adding a &store param, allowing sellers on multi-site setups to specify which store (if any) they'd like the checkout to occur from.
> 
> I have added the noroute behavior in the case where the functionality is disabled. However, I am not comfortable proceeding with the cart deletion logic you have suggested here. This could potentially be a destructive act if the buyer has an active cart session from some other previous shopping session. In the event the Add To Cart Link is invalid, I would prefer not to delete their pre-existing cart.

Hi @sol-loup,

It makes sense to enable multi-store support, because without it, the URL just keeps redirecting to the default store. 

Also thank you for adding the configuration and adding noroute behavior in the case where the functionality is disabled. 
However, in [here](https://github.com/magento/magento2/pull/39667#issuecomment-2785263799), I was mentioning about Request Handling Inconsistencies. If required parameters, **products, coupon, stores** etc. are missing, invalid or invalid URL(tweaked URL), then, the API should return a clear error message. It’d be great to have input validation for parameter keys, so users get clear error messages. Right now, that doesn’t seem to be happening.

Even for the coupon code validation, currently, when an invalid or non-existent coupon code is included in the URL, the coupon appears to be "applied" visually, showing the cancel coupon button without any backend verification. This behavior is misleading and needs to be corrected to ensure only valid coupons trigger any front-end or back-end changes. We need to validate coupon codes before showing that coupon was applied.

Overall,  we need to implement a unified validation to ensure that all requests are consistently handled. 

Can you please look into above points along the with the conflicts resolution in PR.
